src/compiler/preprocess/index.ts

```diff
@@ -76,60 +123,107 @@ export default async function preprocess(
+
+	// TODO keep track: what preprocessor generated what sourcemap? to make debugging easier = detect low-resolution sourcemaps in fn combine_mappings
+
 	for (const fn of markup) {
+
+		// run markup preprocessor
 		const processed = await fn({
 			content: source,
 			filename
 		});
+
 		if (processed && processed.dependencies) dependencies.push(...processed.dependencies);
 		source = processed ? processed.code : source;
+		if (processed && processed.map)
+			sourcemap_list.unshift(
+				typeof(processed.map) === 'string'
+					? JSON.parse(processed.map)
+					: processed.map
+			);
 	}
 
 	for (const fn of script) {
-		source = await replace_async(
+		const get_location = getLocator(source);
+		const res = await replace_async(
+			filename,
 			source,
+			get_location,
 			/<!--[^]*?-->|<script(\s[^]*?)?(?:>([^]*?)<\/script>|\/>)/gi,
-			async (match, attributes = '', content = '') => {
+			async (match, attributes = '', content = '', offset) => {
+				const no_change = () => StringWithSourcemap.from_source(
+					filename, match, get_location(offset));
 				if (!attributes && !content) {
-					return match;
+					return no_change();
 				}
 				attributes = attributes || '';
+				content = content || '';
+
+				// run script preprocessor
 				const processed = await fn({
 					content,
 					attributes: parse_attributes(attributes),
 					filename
 				});
 				if (processed && processed.dependencies) dependencies.push(...processed.dependencies);
-				return processed ? `<script${attributes}>${processed.code}</script>` : match;
+				return processed
+					? get_replacement(filename, offset, get_location, content, processed, `<script${attributes}>`, '</script>')
+					: no_change();
 			}
 		);
+		source = res.string;
+		sourcemap_list.unshift(res.map);
 	}
 
 	for (const fn of style) {
-		source = await replace_async(
+		const get_location = getLocator(source);
+		const res = await replace_async(
+			filename,
 			source,
+			get_location,
 			/<!--[^]*?-->|<style(\s[^]*?)?(?:>([^]*?)<\/style>|\/>)/gi,
-			async (match, attributes = '', content = '') => {
+			async (match, attributes = '', content = '', offset) => {
+				const no_change = () => StringWithSourcemap.from_source(
+					filename, match, get_location(offset));
 				if (!attributes && !content) {
-					return match;
+					return no_change();
 				}
+				attributes = attributes || '';
+				content = content || '';
+
+				// run style preprocessor
 				const processed: Processed = await fn({
 					content,
 					attributes: parse_attributes(attributes),
 					filename
 				});
 				if (processed && processed.dependencies) dependencies.push(...processed.dependencies);
-				return processed ? `<style${attributes}>${processed.code}</style>` : match;
+				return processed
+					? get_replacement(filename, offset, get_location, content, processed, `<style${attributes}>`, '</style>')
+					: no_change();
 			}
 		);
+		source = res.string;
+		sourcemap_list.unshift(res.map);
```

Comment:
Isn't the code inside script and style exactly the same except for the regex? If so, we should extract this into a function

Reply:
yep, can do. The aim here was to keep the line change count small. This change is easier to reason about than one that includes a refactor into its own method (the redundancy already existed)